javalib/src/main/scala/java/util/Base64.scala

```diff
@@ -0,0 +1,486 @@
+                out <<= 8
+                shift = DecodeState14
+                1
+
+              case DecodeState14 =>
+                out |= v << 14
+                b(off+written) = (out >> 16).toByte
+                out <<= 8
+                shift = DecodeState16
+                1
+
+              case DecodeState16 =>
+                out |= v << 16
+                b(off+written) = (out >> 16).toByte
+                out = 0
+                shift = DecodeState18
+                1
+            }
+        }
+      }
+
+      @inline
+      def endOfFile(): Int = {
+        eof = true
+        shift match {
+          case DecodeState18 => 0 // nothing
+          case DecodeState12 => throw new IOException("Base64 stream has one un-decoded dangling byte.")
+          case _             => writeValue(Integer.MAX_VALUE)
+        }
+      }
+
+      @inline
+      def padding(): Int = {
+        eof = true
+        val s = shift
+        if(s == DecodeState18 || s == DecodeState12 || (s == DecodeState14 && in.read() != '=' && !ignoreInvalid))
+          throw new IOException ("Illegal base64 ending sequence")
+        else
+          writeValue(Integer.MAX_VALUE)
+      }
+
+      def iterate(): Unit = {
+        // scalastyle:off return
+        while (written < len)
+          in.read() match {
+            case -1 =>
+              written += endOfFile()
+              return
+
+            case '=' => written += padding()
+            case int => written += writeValue(int)
+          }
+        // scalastyle:on return
+      }
+
+      iterate()
+      written
+    }
+
+    override def close(): Unit = if (!closed) {
+      closed = true
+      in.close()
+    }
+  }
+
+  // --------------------------------------------------------------------------
+
+  class Encoder private[Base64] (table: Array[Byte],
+                                 lineLength: Int = 0,
+                                 lineSeparator: Array[Byte] = Array.empty,
+                                 withPadding: Boolean = true) {
+
+    def encode(src: Array[Byte]): Array[Byte] = {
+      val dst = new Array[Byte](dstLength(src.length))
+      doEncode(src, dst, dst.length)
+      dst
+    }
+
+    def encode(src: Array[Byte], dst: Array[Byte]): Int = {
+      val dstLen = dstLength(src.length)
+      if (dst.length < dstLen)
+        throw new IllegalArgumentException("Output byte array is too small for encoding all input bytes")
+      doEncode(src, dst, dstLen)
+    }
+
+    def encodeToString(src: Array[Byte]): String = new String(encode(src), StandardCharsets.ISO_8859_1)
+
+    def wrap(os: OutputStream): OutputStream =
+      new EncodingOutputStream(os, table, lineLength, lineSeparator, withPadding)
+
+    def encode(buffer: ByteBuffer): ByteBuffer = {
+      val result = new Array[Byte](dstLength(buffer.remaining()))
+      val written =
+        doEncode(new Wrapper(buffer.array(), buffer.position, buffer.limit()), new Wrapper(result))
```

Comment:
Same comment about using `array()`, obviously. tl;dr, I recommend not to. Read into a temporary `Array[Byte]` instead.